Initial Azure Integration

jdunkerley opened the pull request:
### Pull Request Description

- First few method of authentication (Default, CLI, ID/Secret, Environment)
- Ability to list tenants, subscriptions
- Ability to list Storage Accounts
- Ability to list Blob Containers
- Ability to list items in a Blob Container
- Ability to load from a Blob Container
- Ability to load to a Blob Container
 
### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

jdunkerley committed:
Initial Azure connection work

jdunkerley committed:
Initial Blob Storage reading.

jdunkerley committed:
New top level entry for Azure login and identity functions.
Allow secrets to be used in Client Secret auth.

jdunkerley committed:
Doc comment

jdunkerley committed:
Java format.
CHANGELOG entry.

jdunkerley committed:
Adding some error handling.

jdunkerley committed:
Functional list and read of blob.

jdunkerley committed:
Legal review for Azure libs.

jdunkerley committed:
Java format.

hubertp commented:
So.. polyglot deps have increased from ~1.5MB to ~51MB. @Akirathan you need to shrink NI faster than @jdunkerley adds new deps :)

hubertp committed:
Fix NI build

hubertp commented:
If you wonder why this introduces so many dependencies e.g.
```
+-com.azure:azure-core:1.55.3
[info]   | | +-com.azure:azure-json:1.5.0
[info]   | | +-com.azure:azure-xml:1.2.0
[info]   | | +-com.fasterxml.jackson.core:jackson-annotations:2.17.2 (evicted by: 2.1..
[info]   | | +-com.fasterxml.jackson.core:jackson-annotations:2.18.1
[info]   | | +-com.fasterxml.jackson.core:jackson-core:2.17.2 (evicted by: 2.18.1)
[info]   | | +-com.fasterxml.jackson.core:jackson-core:2.18.1
[info]   | | +-com.fasterxml.jackson.core:jackson-databind:2.17.2 (evicted by: 2.18.1)
[info]   | | +-com.fasterxml.jackson.core:jackson-databind:2.18.1
[info]   | | | +-com.fasterxml.jackson.core:jackson-annotations:2.17.2 (evicted by: 2..
[info]   | | | +-com.fasterxml.jackson.core:jackson-core:2.17.2 (evicted by: 2.18.1)
[info]   | | | 
[info]   | | +-com.fasterxml.jackson.datatype:jackson-datatype-jsr310:2.17.2
[info]   | | | +-com.fasterxml.jackson.core:jackson-annotations:2.17.2 (evicted by: 2..
[info]   | | | +-com.fasterxml.jackson.core:jackson-core:2.17.2 (evicted by: 2.18.1)
[info]   | | | +-com.fasterxml.jackson.core:jackson-databind:2.17.2 (evicted by: 2.18..
```
then https://devblogs.microsoft.com/azure-sdk/replacing-jackson-databind-with-azure-json-and-azure-xml/ provides some answers. 
The weird part is that they now not only have jackson as a dependency but their own xml/json libs. On top of that, despite claiming to remove `jackson-databind-xml` completely, failing NI build revealed that not to be entirely true:
```
Caused by: java.lang.IllegalStateException: Failed to retrieve invoker used to create XmlMapper. XML serialization won't be supported until 'com.fasterxml.jackson.dataformat:jackson-dataformat-xml' is added to the classpath or updated to a supported version. Package versions: jackson-core=2.15.2, jackson-databind=2.15.2, jackson-dataformat-xml=unknown, jackson-datatype-jsr310=2.17.2, azure-core=1.55.3, Troubleshooting version conflicts: https://aka.ms/azsdk/java/dependency/troubleshoot
	at com.azure.core.implementation.jackson.XmlMapperFactory.<init>(XmlMapperFactory.java:75)
	at com.azure.core.implementation.jackson.XmlMapperFactory.<clinit>(XmlMapperFactory.java:37)
	... 66 more
Caused by: java.lang.ClassNotFoundException: com.fasterxml.jackson.dataformat.xml.XmlMapper
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:580)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageClassLoader.loadClass(NativeImageClassLoader.java:682)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:543)
	at com.azure.core.implementation.jackson.XmlMapperFactory.<init>(XmlMapperFactory.java:49)
	... 67 more
Caused by: java.lang.ExceptionInInitializerError
	at com.azure.core.implementation.jackson.ObjectMapperFactory.createXmlMapper(ObjectMapperFactory.java:64)
	at com.azure.core.implementation.jackson.ObjectMapperShim.createXmlMapper(ObjectMapperShim.java:84)
	at com.azure.core.util.serializer.JacksonAdapter$GlobalXmlMapper.<clinit>(JacksonAdapter.java:65)
```
Great.

<blockquote><img src="https://devblogs.microsoft.com/azure-sdk/wp-content/uploads/sites/58/2025/03/jackson-java-blog.png" width="48" align="right"><div><img src="https://devblogs.microsoft.com/azure-sdk/wp-content/uploads/sites/58/2024/10/Microsoft-favicon-300x300.jpg" height="14"> Azure SDK Blog</div><div><strong><a href="https://devblogs.microsoft.com/azure-sdk/replacing-jackson-databind-with-azure-json-and-azure-xml/">Replacing jackson-databind with azure-json and azure-xml</a></strong></div><div>jackson-databind (Jackson) is performant, reliable, and universally known in the Java ecosystem. Unfortunately, our dependency on it results in pain points when consuming the Azure SDK client libraries for Java. To reduce and replace Jackson usage, the Azure SDK for Java now offers two new serialization libraries: azure-jsonand azure-xml. These libraries support JSON and XML serialization respectively, providing streaming readers and writers, and interfaces for models to implement serialization behaviors. Over the last few years, the Azure SDK for Java team worked to limit our reliance on jackson-databind.</div></blockquote>

GregoryTravis commented on std-bits/microsoft/src/main/java/org/enso/microsoft/azure/AzureCredential.java:
```diff
@@ -0,0 +1,59 @@
+
+  /**
+   * Represents an Azure credential that uses the environment variables to authenticate. The
+   * required variables are: - `AZURE_TENANT_ID` - `AZURE_CLIENT_ID` - `AZURE_CLIENT_SECRET`
```
```suggestion
   * required variables are: `AZURE_TENANT_ID`, `AZURE_CLIENT_ID`, `AZURE_CLIENT_SECRET`
```

GregoryTravis commented on std-bits/microsoft/src/main/java/org/enso/microsoft/azure/AzureResourceManager.java:
```diff
@@ -0,0 +1,111 @@
+  public record AzureSubscription(String id, String name) {}
+
+  /**
+   * Creates an Azure Resource Manager client using the provided credential and environment.
```
```suggestion
   * Lists Azure subscriptions using the provided credential and environment.
```

hubertp committed:
fmt

hubertp committed:
bump size temporarily

hubertp reviewed (commented):
(only considering new dependencies and how it affects our NI build)
Some problems need to be addressed before this can be merged:
- azure comes with jna `5.13.0`, we are on `5.14.0` and due to its heavy native libraries we definitely do not want to have any duplication here. This has been addressed already twice before
- plenty of netty dependencies that are OS dependent. We need to filter only those which are needed

hubertp committed:
Reduce the number of dependencies

Some dependencies are clearly platform dependent and there is no need to
include them (why include macos stuff on linux?).
Also fixed DependencyFilter that was overlearly eager for organizations.

hubertp committed:
Merge branch 'develop' into wip/jd/azure-blob

hubertp committed:
Update latest native_libs data

hubertp committed:
Bump a few limits to make it at least build NI

hubertp committed:
Bump one more time

Previously it passed with 560. Now it needs one more MB. :shrug:

hubertp committed:
Update API

jdunkerley committed:
PR comments.

AdRiley commented on std-bits/microsoft/src/main/java/org/enso/microsoft/azure/AzureBlobStorage.java:
```diff
@@ -0,0 +1,89 @@
+        result.add(containerName);
+      }
+    }
+    return result;
```
Stylistic suggestion. Feel free to ignore.

```suggestion
    return containerList.stream()
        .map(BlobContainerItem::getName)
        .filter(name -> prefix.isEmpty() || name.startsWith(prefix))
        .collect(Collectors.toList());
```

JaroslavTulach commented on distribution/lib/Standard/Microsoft/0.0.0-dev/THIRD-PARTY/NOTICE:
- More than 200 of new license files. 
- How long did it take to process them?
- Was it worth the effort?

jdunkerley replied:
About 45 minutes. Its the process we have at the moment and no I don't love it.

JaroslavTulach commented on distribution/lib/Standard/Microsoft/0.0.0-dev/src/Main.enso:
```diff
@@ -1,2 +1,4 @@
 export project.Connection.SQLServer_Details.SQLServer_Details
 
+export project.Azure.Azure
+export project.Azure_Storage.Azure_Storage
```
This is the import **important API change**. When people `from Standard.Microsoft import all` they will now see `Azure` and `Azure_Storage`.

@Akirathan, I am surprised this change isn't reflected in the `docs/api` at all! Shouldn't we be tracking also _exported symbols_?

JaroslavTulach commented on distribution/lib/Standard/Microsoft/0.0.0-dev/src/Azure_Credential.enso:
```diff
@@ -0,0 +1,101 @@
+## PRIVATE
+Text.from (that:Azure_Subscription) -> Text = that.id
+
+## PRIVATE
```
There is no point of adding `## PRIVATE` to a `private` method.

jdunkerley replied:
Used by my doc rewriter at the moment.

jdunkerley replied:
Amended the ensolint script to ignore these - so wont be added in future.
Also covered the conversion methods so they are ignored if not doc commented.

JaroslavTulach commented on distribution/lib/Standard/Microsoft/0.0.0-dev/src/Azure_Credential.enso:
```diff
@@ -0,0 +1,101 @@
+        name = ["name", self.name]
+        JS_Object.from_pairs [["type", "Azure_Subscription"], id, name]
+
+## PRIVATE
```
I don't thing conversions methods are displayed anywhere by the IDE. Consider removing the `## PRIVATE` comment - it is useless.

JaroslavTulach commented on distribution/lib/Standard/Microsoft/0.0.0-dev/src/Azure_Credential.enso:
```diff
@@ -0,0 +1,101 @@
+
+## Represents an Azure subscription
+type Azure_Subscription
+    private Value java_object
```
+1 to encapsulation of the constructors

JaroslavTulach commented on distribution/lib/Standard/Microsoft/0.0.0-dev/src/Azure_Credential.enso:
```diff
@@ -0,0 +1,101 @@
+
+## Represents an Azure tenant
+type Azure_Tenant
+    Value id
```
- Why isn't this constructor encapsulated?
- Shouldn't the `id` type (or type of `id` getter if the constructor is made `private`) be specified?

JaroslavTulach commented on distribution/lib/Standard/Microsoft/0.0.0-dev/src/Azure_Credential.enso:
```diff
@@ -0,0 +1,101 @@
+    CLI
+
+    ## Read the credential from inputs or Enso Secrets
+    Client_Secret (tenant_id : (Text|Enso_Secret) = (Missing_Argument.throw "tenant_id")) (client_id : (Text|Enso_Secret) = (Missing_Argument.throw "client_id")) (client_secret : (Text|Enso_Secret) = (Missing_Argument.throw "client_secret"))
```
- exposing constructors to public restricts the ability to modify # of constructors in the future compatibly
- should there be `case azul_credential of` code covering all cases, adding new constructor would  break it

jdunkerley replied:
This is a standard Enso pattern allowing the user to pick the authentication method. The widget displays the set of constructors and the user chooses.

Yes, we have to cover these cases - ideally we would have some static analysis to detect if we missed such things.

JaroslavTulach commented on distribution/lib/Standard/Microsoft/0.0.0-dev/src/Azure_Credential.enso:
```diff
@@ -0,0 +1,101 @@
+        JS_Object.from_pairs [["type", "Azure_Subscription"], id, name]
+
+## PRIVATE
+Text.from (that:Azure_Subscription) -> Text = that.id
```
- a conversion `.to Text`!
- where is such a conversion used/useful?
- there were issues requesting usage of `.to Text`, but...
- as far as I know we are still relying on `.to_text`, `.to_display_text`, etc.

jdunkerley replied:
It allows passing it into a method with the value being `Text`.

jdunkerley replied:
Yes we still rely on to_text and to_display_text

JaroslavTulach commented on distribution/lib/Standard/Microsoft/0.0.0-dev/src/Azure_Storage.enso:
```diff
@@ -0,0 +1,64 @@
+
+## Enso representation of the Azure Storage Account
+type Azure_Storage_Account
+    private Value java_object
```
+1

JaroslavTulach commented on project/GraalVM.scala:
```diff
@@ -99,11 +99,11 @@ object GraalVM {
-    private val macX64Release     = NativeImageSize(200, 408)
+    private val windowsX64Release = NativeImageSize(200, 440)
+    private val linuxX64Release   = NativeImageSize(200, 465)
+    private val macX64Release     = NativeImageSize(200, 426)
```
- Unfortunate from a technology perspective
- However, when there is a business case, architecture has to wait
- at least another incentive to work on #13172

JaroslavTulach commented on std-bits/base/src/main/java/org/enso/base/enso_cloud/ExternalLibrarySecretHelper.java:
```diff
@@ -15,6 +15,7 @@ public static String resolveValue(HideableValue hideableValue) throws EnsoSecret
+          new RestrictedAccess.AccessLocation("org.enso.aws.ClientBuilder", "unsafeResolveSecrets"),
           new RestrictedAccess.AccessLocation(
-              "org.enso.aws.ClientBuilder", "unsafeResolveSecrets"));
+              "org.enso.microsoft.azure.CredentialHelper", "unsafeResolveSecrets"));
```
- another motivation for #11724
- why is the method called `unsafeResolveSecrets`?

jdunkerley replied:
It has nothing to do with SPI loading.
We choose to whitelist methods that are allowed to read the raw secret value. The method is unsafe as we do not want the value to end up in the Enso code.

JaroslavTulach commented on std-bits/microsoft/src/main/java/org/enso/microsoft/azure/LRUCache.java:
```diff
@@ -0,0 +1,26 @@
+ * @param <K> the type of keys
+ * @param <V> the type of values
+ */
+class LRUCache<K, V> extends java.util.LinkedHashMap<K, V> {
```
```suggestion
final class LRUCache<K, V> extends java.util.LinkedHashMap<K, V> {
```

JaroslavTulach commented on test/Base_Tests/data/native_libs.json:
```diff
@@ -147,6 +147,36 @@
     "META-INF/native/libconscrypt_openjdk_jni-osx-x86_64.dylib"
   ],
+  "lib/Standard/Microsoft/0.0.0-dev/polyglot/java/jna-wrapper-assembly-0.1.0-SNAPSHOT.jar": [
+    "com/sun/jna/freebsd-x86-64/libjnidispatch.so",
```
- Why do we need another JNA in the system?
- I believe these `.so` & co. files are already present somewhere
- Can't we just re-use them?

hubertp replied:
We do not. If you read `build.sbt` changes again I'm explicitly excluding jna that is part of `azure-identity`. `jna-wrapper-assembly` is already part of NI so it makes no difference to the overall size. It has not been part of any of stdlibs dependencies.

JaroslavTulach replied:
> jna-wrapper-assembly is already part of NI so it makes no difference to the overall size

I don't understand. How can a new JAR at `lib/Standard/Microsoft/0.0.0-dev/polyglot/java/jna-wrapper-assembly-0.1.0-SNAPSHOT.jar` make _no difference to overall size_? 

- Why do we need another JNA JAR in the system?
- How can you claim a new JAR in the system makes _no difference to overall size_?

JaroslavTulach commented on test/Base_Tests/data/native_libs.json:
```diff
@@ -147,6 +147,36 @@
+    "META-INF/native/libnetty_tcnative_linux_x86_64.so"
+  ],
+  "lib/Standard/Microsoft/0.0.0-dev/polyglot/java/netty-transport-native-epoll-4.1.118.Final-linux-x86_64.jar": [
+    "META-INF/native/libnetty_transport_native_epoll_x86_64.so"
```
- I see `.so` files for netty
- where is the `.dll`?
- isn't this functionality optional? 
- if the Java code can run without a native library on Windows - can't it also run without a native library on Linux?

hubertp replied:
Java code can't run without a native library on Windows. See [header](https://github.com/enso-org/enso/blob/develop/test/Base_Tests/src/System/Embedded_Native_Libs_Spec.enso#L10) of the test file for explanation - test runs only on Linux.

Akirathan replied:
Note that https://github.com/enso-org/enso/pull/13174 will probably be merged sooner than this PR. So we will need to deal with native libraries on Windows as well.

jdunkerley replied:
As this is set to merge shortly as well what are the implications?

Akirathan replied:
I have just merged that PR. I am not entirely sure. We will have to see the specific errors in tests now.

JaroslavTulach replied:
> ...implications...

- #13193

hubertp committed:
Removing one additional dependency

Simplified dependency exclusion. Rather than excluding them at the
copying level, excluded jna at the dependency definition site.
`netty-resolver-dns-classes-macos` needs to be OS-dependent rather than
being always excluded.

jdunkerley committed:
Java fmt and PR comments.

jdunkerley committed:
Legal Review
